### PR TITLE
Add FlappyDon as a second game template

### PR DESCRIPTION
**Pull request by TimOliver**
Adds a fully featured, and playable second game template, based on Flappy Bird to the osu-framework sample game pool.

### CODE REVIEW

**Comment by peppy** [comment]
Took the liberty of rebasing this away from your other template PR for review purposes.

File: osu.Framework.Templates/templates/template-flappy/FlappyDon.iOS/FlappyDonIOS.cs [hunk 1]
```diff
@@ -0,0 +1,11 @@
+﻿// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using FlappyDon.Game;
+
+namespace FlappyDon.iOS
+{
+    public class FlappyDonIOS : FlappyDonGame
+    {
+    }
```
[the remark this answers is not in the document]
Comment (TimOliver) [reply osu.Framework.Templates/templates/template-flappy/FlappyDon.iOS/FlappyDonIOS.cs]: Done!

File: osu.Framework.Templates/templates/template-flappy/FlappyDon.Game.Tests/Visual/TestSceneMyNewProjectGame.cs [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+
+using System;
+using System.Collections.Generic;
+using osu.Framework.Allocation;
+using osu.Framework.Platform;
+using osu.Framework.Testing;
+
+namespace FlappyDon.Game.Tests.Visual
+{
+    public class TestSceneFlappyDonGame : TestScene
```
[the remark this answers is not in the document]
Comment (TimOliver) [reply osu.Framework.Templates/templates/template-flappy/FlappyDon.Game.Tests/Visual/TestSceneMyNewProjectGame.cs]: Done!

File: osu.Framework.Templates/templates/template-flappy/FlappyDon.Game/Backdrop/Backdrop.cs [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+        // used to determine when we need to
+        // perform a fresh layout
+        private Vector2 lastSize;
+
+        // Whether the sprite set is currently
+        // animating or not
+        public bool Running { get; private set; }
+
+        // The duration it takes to animate one cycle
+        // of the sprintes in the container.
```
[the remark this answers is not in the document]
Comment (TimOliver) [reply osu.Framework.Templates/templates/template-flappy/FlappyDon.Game/Backdrop/Backdrop.cs]: Nicely spotted! I updated all of the typos!

File: osu.Framework.Templates/templates/template-flappy/FlappyDon.Game/Backdrop/Backdrop.cs [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+                // Change the inset to overlap slightly in order to avoid potential seam lines
+                offset += width - 1;
+            }
+        }
+
+        private void stopAnimatingChildren()
+        {
+            foreach (var childSprite in Children)
+                childSprite.ClearTransforms();
+        }
```
[the remark this answers is not in the document]
Comment (TimOliver) [reply osu.Framework.Templates/templates/template-flappy/FlappyDon.Game/Backdrop/Backdrop.cs]: 👍

File: osu.Framework.Templates/templates/template-flappy/FlappyDon.Game/FullScreen/ScoreSpriteText.cs [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+using osu.Framework.Text;
+
+namespace FlappyDon.Game
+{
+    /*
+     * Draws the current score on screen.
+     * The score is displayed by using a glyph store
+     * to load the appropriate character images and
+     * display them like text on-screen.
+     */
```
[the remark this answers is not in the document]
Comment (TimOliver) [reply osu.Framework.Templates/templates/template-flappy/FlappyDon.Game/FullScreen/ScoreSpriteText.cs]: Oops. That was the formatting style I was using before. I must have missed that one. Thanks!

File: osu.Framework.Templates/templates/template-flappy/FlappyDon.Game/Testing/MyNewProjectTestBrowser.cs [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+
+using osu.Framework.Graphics;
+using osu.Framework.Graphics.Containers;
+using osu.Framework.Graphics.Cursor;
+using osu.Framework.Platform;
+using osu.Framework.Testing;
+
+namespace FlappyDon.Game.Testing
+{
+    public class FlappyDonTestBrowser : FlappyDonGame
```
[the remark this answers is not in the document]
Comment (TimOliver) [reply osu.Framework.Templates/templates/template-flappy/FlappyDon.Game/Testing/MyNewProjectTestBrowser.cs]: Thanks for the clarification! I worked this one out with peppy. Hopefully it should be all good now.

File: osu.Framework.Templates/templates/template-flappy/FlappyDon.Desktop/Program.cs [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+using osu.Framework.Platform;
+
+namespace FlappyDon.Desktop
+{
+    public static class Program
+    {
+        public static void Main()
+        {
+            using (GameHost host = Host.GetSuitableHost(@"FlappyDon"))
+            using (osu.Framework.Game game = new FlappyDonGameBase())
```
[the remark this answers is not in the document]
Comment (TimOliver) [reply osu.Framework.Templates/templates/template-flappy/FlappyDon.Desktop/Program.cs]: Thanks for the heads up! I went back and refactored the whole base game class hierarchy (And their subsequent test scenes), so HOPEFULLY this is now resolved in a better way.

**Comment by TimOliver** [comment]
Alrighty. Sorry for the delay. I've gone back and refactored the game class objects (As well as set back up the test browser build targets). Hopefully this should all be good now. :)

**Comment by smoogipoo** [comment]
Need to fix the pipes going above (and potentially below) the screen and the bird going below the ground on different resolutions.

**Comment by TimOliver** [comment]
@smoogipoo Okay! I've confirmed the new scaling hasn't distorted anything. Should be good to go. Let me know if you need me to do anything else! :)

File: osu.Framework.Templates/templates/template-flappy/FlappyDon.Game/FlappyDonGame.cs [hunk 8]
```diff
@@ -0,0 +1,236 @@
...
+
+            if (gameOver)
+            {
+                reset();
+                return base.OnMouseDown(e);
+            }
+
+            onTapEvent();
+
+            return base.OnMouseDown(e);
```
[the remark this answers is not in the document]
Comment (TimOliver) [reply osu.Framework.Templates/templates/template-flappy/FlappyDon.Game/FlappyDonGame.cs]: Good call! Done!

File: osu.Framework.Templates/templates/template-flappy/FlappyDon.Game/FlappyDonGame.cs [hunk 9]
```diff
@@ -0,0 +1,236 @@
...
+            if (gameOver)
+            {
+                reset();
+                return base.OnKeyDown(e);
+            }
+
+            if (e.Key == Key.Space && e.Repeat == false)
+                onTapEvent();
+
+            return base.OnKeyDown(e);
```
[the remark this answers is not in the document]
Comment (TimOliver) [reply osu.Framework.Templates/templates/template-flappy/FlappyDon.Game/FlappyDonGame.cs]: Done!

File: osu.Framework.Templates/templates/template-flappy/FlappyDon.Game.Tests/Visual/TestSceneFlappyDonGame.cs [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+    /// A test scene wrapping the entire game,
+    /// including audio.
+    /// </summary>
+    public class TestSceneFlappyDonGame : TestScene
+    {
+        private FlappyDonGame game;
+
+        public override IReadOnlyList<Type> RequiredTypes => new[]
+        {
+            typeof(FlappyDonGame),
```
[the remark this answers is not in the document]
Comment (TimOliver) [reply osu.Framework.Templates/templates/template-flappy/FlappyDon.Game.Tests/Visual/TestSceneFlappyDonGame.cs]: Removed.

File: osu.Framework.Templates/templates/template-flappy/FlappyDon.Game/Elements/Backdrop.cs [hunk 11]
```diff
@@ -0,0 +1,141 @@
...
+using osu.Framework.Allocation;
+using osu.Framework.Graphics;
+using osu.Framework.Graphics.Containers;
+using osu.Framework.Graphics.Sprites;
+using osuTK;
+
+namespace FlappyDon.Game.Elements
+{
+    /// <summary>
+    /// Manages a pool of Sprite objects,
```
[the remark this answers is not in the document]
Comment (TimOliver) [reply osu.Framework.Templates/templates/template-flappy/FlappyDon.Game/Elements/Backdrop.cs]: Aha. I got into this habit after hearing a lot of tech companies still enforce a very strict "80-characters per line" policy. I'll keep that in mind for here.

I've gone through and updated as many as I can.

File: osu.Framework.Templates/templates/template-flappy/FlappyDon.Game/Elements/Backdrop.cs [hunk 12]
```diff
@@ -0,0 +1,141 @@
...
+    /// arranged and animated horizontally
+    /// to produce a background scrolling effect.
+    /// </summary>
+    public class Backdrop : CompositeDrawable
+    {
+        // Holds a lambda that generates new
```
[the remark this answers is not in the document]
Comment (TimOliver) [reply osu.Framework.Templates/templates/template-flappy/FlappyDon.Game/Elements/Backdrop.cs]: Fair enough! I went back and updated the documentation for all of the public methods. Hope that helps! :)

**Comment by peppy** [comment]
I've done some further cleanup on comments and general style, but as a template project I think we need to go through the actual game logic and fix things up that are hard to understand in the current implementation (as a sample game it's probably fine, but as a template I believe the current level of code quality would not set a user up in a good state to extend on the game).

File: osu.Framework.Templates/templates/template-flappy/FlappyDon.Game/Elements/Backdrop.cs [hunk 13]
```diff
@@ -0,0 +1,147 @@
...
+
+        /// <summary>
+        /// Whether the sprite set is currently animating or not.
+        /// </summary>
+        public bool Running { get; private set; }
+
+        /// <summary>
+        /// The duration it takes to animate one cycle of the sprites in the container.
+        /// </summary>
+        public float Duration;
```
Comment (peppy) [inline osu.Framework.Templates/templates/template-flappy/FlappyDon.Game/Elements/Backdrop.cs]: shouldn't be public settable. should probably be `readonly`.

File: osu.Framework.Templates/templates/template-flappy/FlappyDon.Game/Elements/Backdrop.cs [hunk 14]
```diff
@@ -0,0 +1,147 @@
...
+            Running = false;
+            stopAnimatingChildren();
+        }
+
+        protected override void UpdateAfterChildren()
+        {
+            base.UpdateAfterChildren();
+
+            // If the bounds of this container changed, add or remove the number of child sprites to fill the visible space.
+            if (DrawSize.Equals(lastSize) == false)
```
Comment (peppy) [inline osu.Framework.Templates/templates/template-flappy/FlappyDon.Game/Elements/Backdrop.cs]:
```suggestion
            if (!DrawSize.Equals(lastSize)z)
```

File: osu.Framework.Templates/templates/template-flappy/FlappyDon.Game/Elements/Backdrop.cs [hunk 15]
```diff
@@ -0,0 +1,147 @@
...
+            }
+        }
+
+        protected override void LoadComplete()
+        {
+            base.LoadComplete();
+            updateChildrenList();
+        }
+
+        private void updateChildrenList()
```
Comment (peppy) [inline osu.Framework.Templates/templates/template-flappy/FlappyDon.Game/Elements/Backdrop.cs]: rename to `updateLayout`

File: osu.Framework.Templates/templates/template-flappy/FlappyDon.Game/Elements/Backdrop.cs [hunk 16]
```diff
@@ -0,0 +1,147 @@
...
+            {
+                updateChildrenList();
+                lastSize = DrawSize;
+            }
+        }
+
+        protected override void LoadComplete()
+        {
+            base.LoadComplete();
+            updateChildrenList();
```
Comment (peppy) [inline osu.Framework.Templates/templates/template-flappy/FlappyDon.Game/Elements/Backdrop.cs]: unnecessary; it will be run in `UpdateAfterChildren`.

File: osu.Framework.Templates/templates/template-flappy/FlappyDon.Game/Elements/Backdrop.cs [hunk 17]
```diff
@@ -0,0 +1,147 @@
...
+        {
+            // Get an initial sprite we can use to derive size
+            Sprite sprite;
+
+            // Fetch an initial child sprite we can use to measure
+            if (InternalChildren.Count > 0)
+                sprite = (Sprite)InternalChildren.First();
+            else
+            {
+                // TODO: this is probably wrong. we are adding the first sprite *twice* – once here and once below.
```
Comment (peppy) [inline osu.Framework.Templates/templates/template-flappy/FlappyDon.Game/Elements/Backdrop.cs]: weird logic here. needs a re-think.

File: osu.Framework.Templates/templates/template-flappy/FlappyDon.Game/Elements/BackdropSprite.cs [hunk 18]
```diff
@@ -0,0 +1,53 @@
...
+namespace FlappyDon.Game.Elements
+{
+    /// <summary>
+    /// Manages the buildings and sky texture in the far background of the game scene.
+    /// The texture is aspect scaled up to the height of the game window.
+    /// </summary>
+    public class BackdropSprite : Sprite
+    {
+        [Resolved]
+        private TextureStore textures { get; set; }
```
Comment (peppy) [inline osu.Framework.Templates/templates/template-flappy/FlappyDon.Game/Elements/BackdropSprite.cs]: move to parameter (only used in `load()`)

File: osu.Framework.Templates/templates/template-flappy/FlappyDon.Game/Elements/BackdropSprite.cs [hunk 19]
```diff
@@ -0,0 +1,53 @@
...
+
+        public float AspectRatio
+        {
+            get
+            {
+                if (Texture == null) return 1.0f;
```
Comment (peppy) [inline osu.Framework.Templates/templates/template-flappy/FlappyDon.Game/Elements/BackdropSprite.cs]: when is `Texture` null?

File: osu.Framework.Templates/templates/template-flappy/FlappyDon.Game/Elements/BackdropSprite.cs [hunk 20]
```diff
@@ -0,0 +1,53 @@
...
+            Anchor = Anchor.TopLeft;
+            Origin = Anchor.TopLeft;
+        }
+
+        [BackgroundDependencyLoader]
+        private void load()
+        {
+            Texture = textures.Get("background-day");
+            RelativeSizeAxes = Axes.Y;
+            Size = new Vector2(320.0f, 1.0f);
```
Comment (peppy) [inline osu.Framework.Templates/templates/template-flappy/FlappyDon.Game/Elements/BackdropSprite.cs]: might be autosize x instead

File: osu.Framework.Templates/templates/template-flappy/FlappyDon.Game/Elements/BackdropSprite.cs [hunk 21]
```diff
@@ -0,0 +1,53 @@
...
+    /// <summary>
+    /// Manages the buildings and sky texture in the far background of the game scene.
+    /// The texture is aspect scaled up to the height of the game window.
+    /// </summary>
+    public class BackdropSprite : Sprite
+    {
+        [Resolved]
+        private TextureStore textures { get; set; }
+
+        public float AspectRatio
```
Comment (peppy) [inline osu.Framework.Templates/templates/template-flappy/FlappyDon.Game/Elements/BackdropSprite.cs]: move this inline. it's only used in one place and only in a private context

File: osu.Framework.Templates/templates/template-flappy/FlappyDon.Game/Elements/Bird.cs [hunk 22]
```diff
@@ -0,0 +1,138 @@
...
+        public bool IsTouchingGround { get; private set; }
+
+        private readonly TextureAnimation animation = new TextureAnimation();
+
+        [Resolved]
+        private TextureStore textures { get; set; }
+
+        private DrawableSample flapSound;
+
+        private bool isPlaying;
```
Comment (peppy) [inline osu.Framework.Templates/templates/template-flappy/FlappyDon.Game/Elements/Bird.cs]: this state is quite hard to follow. also weird the bird has its own "playing" state

File: osu.Framework.Templates/templates/template-flappy/FlappyDon.Game/Elements/Bird.cs [hunk 23]
```diff
@@ -0,0 +1,138 @@
...
+        private float currentVelocity;
+
+        [BackgroundDependencyLoader]
+        private void load(ISampleStore samples)
+        {
+            Anchor = Anchor.CentreLeft;
+            Origin = Anchor.Centre;
+            Position = new Vector2(120.0f, .0f);
+
+            animation.AddFrame(textures.Get("redbird-upflap"), 200.0f);
```
Comment (peppy) [inline osu.Framework.Templates/templates/template-flappy/FlappyDon.Game/Elements/Bird.cs]: move construction inline and use object initializer

File: osu.Framework.Templates/templates/template-flappy/FlappyDon.Game/Elements/Bird.cs [hunk 24]
```diff
@@ -0,0 +1,138 @@
...
+            animation.AddFrame(textures.Get("redbird-downflap"), 200.0f);
+            animation.AddFrame(textures.Get("redbird-midflap"), 200.0f);
+            animation.Origin = Anchor.Centre;
+            animation.Anchor = Anchor.Centre;
+            animation.Scale = new Vector2(0.5f);
+
+            AddInternal(animation);
+            AddInternal(flapSound = new DrawableSample(samples.Get("wing.ogg")));
+
+            Size = new Vector2(50.0f, 35.0f);
```
Comment (peppy) [inline osu.Framework.Templates/templates/template-flappy/FlappyDon.Game/Elements/Bird.cs]: arbitrary? can we autosize this from the texture?

File: osu.Framework.Templates/templates/template-flappy/FlappyDon.Game/Elements/GroundSprite.cs [hunk 25]
```diff
@@ -0,0 +1,31 @@
...
+
+namespace FlappyDon.Game.Elements
+{
+    /// <summary>
+    /// Manages the ground texture in the foreground of the game scene.
+    /// </summary>
+    public class GroundSprite : Sprite
+    {
+        [Resolved]
+        private TextureStore textures { get; set; }
```
Comment (peppy) [inline osu.Framework.Templates/templates/template-flappy/FlappyDon.Game/Elements/GroundSprite.cs]: move to parameter

File: osu.Framework.Templates/templates/template-flappy/FlappyDon.Game/Elements/Obstacles.cs [hunk 26]
```diff
@@ -0,0 +1,161 @@
...
+using osu.Framework.Utils;
+using osuTK;
+
+namespace FlappyDon.Game.Elements
+{
+    /// <summary>
+    /// A full-screen container used to manage a pool of pipe obstacles. De-queues pipes when they go off screen and creates
+    /// new instances as needed. Also handles collision detection between any pipes and the bird sprite.
+    /// </summary>
+    public class Obstacles : CompositeDrawable
```
Comment (peppy) [inline osu.Framework.Templates/templates/template-flappy/FlappyDon.Game/Elements/Obstacles.cs]: missing test scene. this class has quite complex logic so i'd want to be able to test it out.

File: osu.Framework.Templates/templates/template-flappy/FlappyDon.Game/FlappyDonGame.cs [hunk 27]
```diff
@@ -0,0 +1,224 @@
...
+            {
+                gameOver = true;
+                fail();
+                return;
+            }
+
+            // If a pipe crosses the bird's X offset,
+            // (And no collisions occurred),
+            // increment the score by 1
+            if (obstacles.ThresholdCrossed())
```
Comment (peppy) [inline osu.Framework.Templates/templates/template-flappy/FlappyDon.Game/FlappyDonGame.cs]: this feels really unsafe and backwards. obstacles should probably fire an event when the threshold is crossed

File: osu.Framework.Templates/templates/template-flappy/FlappyDon.Game/FlappyDonGame.cs [hunk 28]
```diff
@@ -0,0 +1,224 @@
...
+        private Backdrop groundBackdrop;
+
+        // Sound effects
+        private DrawableSample scoreSound;
+        private DrawableSample punchSound;
+        private DrawableSample fallSound;
+        private DrawableSample whooshSound;
+
+        // Game state
+        private int score;
```
Comment (peppy) [inline osu.Framework.Templates/templates/template-flappy/FlappyDon.Game/FlappyDonGame.cs]: there should be a score counter component, rather than having an int and spritetext local to the game.

File: osu.Framework.Templates/templates/template-flappy/FlappyDon.Game/FlappyDonGame.cs [hunk 29]
```diff
@@ -0,0 +1,224 @@
...
+            scoreSpriteText.Show();
+
+            // Animate the bird flying up
+            bird.FlyUp();
+        }
+
+        protected override bool OnKeyDown(KeyDownEvent e)
+        {
+            // After dying, disable input briefly to stop the user restarting the game too quickly.
+            if (gameOverCooldown)
```
Comment (peppy) [inline osu.Framework.Templates/templates/template-flappy/FlappyDon.Game/FlappyDonGame.cs]: this logic is copied and pasted from `OnMouseDown`? it should be completely shared, IMO.

File: osu.Framework.Templates/templates/template-flappy/FlappyDon.Game/Elements/Obstacles.cs [hunk 30]
```diff
@@ -0,0 +1,161 @@
...
+
+            Running = false;
+            frozen = false;
+
+            // Remove all child pipes
+            // from this container
+            ClearInternal();
+        }
+
+        public bool CollisionDetected(Quad birdQuad)
```
Comment (peppy) [inline osu.Framework.Templates/templates/template-flappy/FlappyDon.Game/Elements/Obstacles.cs]:
```suggestion
        public bool CheckForCollision(Quad birdQuad)
```

File: osu.Framework.Templates/templates/template-flappy/FlappyDon.Game/Elements/Obstacles.cs [hunk 31]
```diff
@@ -0,0 +1,161 @@
...
+        {
+            if (InternalChildren.Count == 0)
+                return false;
+
+            var obstacle = (PipeObstacle)InternalChildren.First();
+            return obstacle.Colliding(birdQuad);
```
Comment (peppy) [inline osu.Framework.Templates/templates/template-flappy/FlappyDon.Game/Elements/Obstacles.cs]:
```suggestion
=> InternalChildren.FirstOrDefault()?.Cast<PipeObstacle>().Colliding(birdQuad) ?? false;
```

File: osu.Framework.Templates/templates/template-flappy/FlappyDon.Game/Elements/Obstacles.cs [hunk 32]
```diff
@@ -0,0 +1,161 @@
...
+            else
+                spawnNewObstacle();
+        }
+
+        private void spawnNewObstacle()
+        {
+            var obstacle = new PipeObstacle
+            {
+                Position = new Vector2(DrawWidth, 0.0f),
+                Offset = RNG.NextSingle(-140.0f, 60.0f)
```
Comment (peppy) [inline osu.Framework.Templates/templates/template-flappy/FlappyDon.Game/Elements/Obstacles.cs]: bit of an ambiguous variable. maybe `VerticalPositionAdjust`

File: osu.Framework.Templates/templates/template-flappy/FlappyDon.Game/Elements/Pipe.cs [hunk 33]
```diff
@@ -0,0 +1,31 @@
...
+namespace FlappyDon.Game.Elements
+{
+    /// <summary>
+    /// A sprite that shows a single green pipe. There are two pipe sprites for every obstacle the
+    /// player must overcome.
+    /// </summary>
+    public class Pipe : Sprite
+    {
+        [Resolved]
+        private TextureStore textures { get; set; }
```
Comment (peppy) [inline osu.Framework.Templates/templates/template-flappy/FlappyDon.Game/Elements/Pipe.cs]: parameter

File: osu.Framework.Templates/templates/template-flappy/FlappyDon.Game/Elements/PipeObstacle.cs [hunk 34]
```diff
@@ -0,0 +1,84 @@
...
+    /// A single obstacle for the player to overcome, consisting of two pipe sprites, one rotated 180
+    /// degrees and placed along the top of the screen.
+    /// </summary>
+    public class PipeObstacle : CompositeDrawable
+    {
+        /// <summary>
+        /// When the bird has been detected to have crossed the pipe, and the score has been incremented, this flag is set to
+        /// true to denote it has already been counted in the score.
+        /// </summary>
+        public bool Scored = false;
```
Comment (peppy) [inline osu.Framework.Templates/templates/template-flappy/FlappyDon.Game/Elements/PipeObstacle.cs]: Generally we try to avoid having parents setting on fields like this. It leads to unmaintainable code. Either PipeObstacle should be responsible for updating this or the tracking logic should completely be moved to a parent.

File: osu.Framework.Templates/templates/template-flappy/FlappyDon.Game/Elements/PipeObstacle.cs [hunk 35]
```diff
@@ -0,0 +1,84 @@
...
+            {
+                Anchor = Anchor.Centre,
+                Origin = Anchor.TopCentre,
+                Position = new Vector2(0.0f, 110 + Offset)
+            };
+
+            AddInternal(bottomPipe);
+        }
+
+        public bool Colliding(Quad birdQuad)
```
Comment (peppy) [inline osu.Framework.Templates/templates/template-flappy/FlappyDon.Game/Elements/PipeObstacle.cs]: `CheckCollision`

File: osu.Framework.Templates/templates/template-flappy/FlappyDon.Game/Elements/ScreenFlash.cs [hunk 36]
```diff
@@ -0,0 +1,34 @@
...
+
+namespace FlappyDon.Game.Elements
+{
+    /// <summary>
+    /// A full-screen white flash used to indicate a collision with a pipe, or when the game scene resets after a game over.
+    /// </summary>
+    public class ScreenFlash : Box
+    {
+        /// <summary>
+        /// The duration for the screen flash to play.
```
Comment (peppy) [inline osu.Framework.Templates/templates/template-flappy/FlappyDon.Game/Elements/ScreenFlash.cs]: incorrect

File: osu.Framework.Templates/templates/template-flappy/FlappyDon.Game/FlappyDonGame.cs [hunk 37]
```diff
@@ -0,0 +1,224 @@
...
+        {
+            // Play a brief flash to make the hit very visible
+            screenFlash.GameOverFlash();
+
+            // Play the punch sound, and then the 'fall' sound slightly after
+            punchSound.Play();
+            Scheduler.AddDelayed(() => fallSound.Play(), 70.0);
+
+            // Show the game over title in the middle of the flash
+            gameOverSprite.Show(screenFlash.FlashDuration);
```
Comment (peppy) [inline osu.Framework.Templates/templates/template-flappy/FlappyDon.Game/FlappyDonGame.cs]: just do the delay local here. remove it from the `Show` parameters.

**Comment by peppy** [comment]
Just an initial pass. Probably want to look at maintaining current game state as an enum at very least, and tidy things up from that angle too.

```
enum GameState
{
    Menu,
    Playing,
    Failed
}
```

File: osu.Framework.Templates/templates/template-flappy/FlappyDon.Game/Elements/Backdrop.cs (hunk 13 above)
Comment (smoogipoo) [reply osu.Framework.Templates/templates/template-flappy/FlappyDon.Game/Elements/Backdrop.cs]: For time values we also typically work with `double`s.

File: osu.Framework.Templates/templates/template-flappy/FlappyDon.Game/Elements/Backdrop.cs (hunk 17 above)
Comment (smoogipoo) [reply osu.Framework.Templates/templates/template-flappy/FlappyDon.Game/Elements/Backdrop.cs]: As discussed, you should be able to load it once in `load()` just to get the draw width, and then use that in the while loops below. From what it looked like to me, this should all still handle fine with that change.

File: osu.Framework.Templates/templates/template-flappy/FlappyDon.Game/Elements/Backdrop.cs (hunk 13 above)
Comment (TimOliver) [reply osu.Framework.Templates/templates/template-flappy/FlappyDon.Game/Elements/Backdrop.cs]: Done!

File: osu.Framework.Templates/templates/template-flappy/FlappyDon.Game/Elements/Backdrop.cs (hunk 14 above)
Comment (TimOliver) [reply osu.Framework.Templates/templates/template-flappy/FlappyDon.Game/Elements/Backdrop.cs]: Fair enough. Changed.

File: osu.Framework.Templates/templates/template-flappy/FlappyDon.Game/Elements/Backdrop.cs (hunk 17 above)
Comment (TimOliver) [reply osu.Framework.Templates/templates/template-flappy/FlappyDon.Game/Elements/Backdrop.cs]: Done like smoogi said. I've added an initial sprite in the `load` method and removed that conditional

File: osu.Framework.Templates/templates/template-flappy/FlappyDon.Game/Elements/BackdropSprite.cs (hunk 20 above)
Comment (TimOliver) [reply osu.Framework.Templates/templates/template-flappy/FlappyDon.Game/Elements/BackdropSprite.cs]: The value doesn't seem to matter, as long as it's not 0.
To avoid hard-coding values, I changed it to `Texture.Size.Width`

File: osu.Framework.Templates/templates/template-flappy/FlappyDon.Game/Elements/Bird.cs (hunk 22 above)
Comment (TimOliver) [reply osu.Framework.Templates/templates/template-flappy/FlappyDon.Game/Elements/Bird.cs]: It's supposed to denote when the player is actively controlling the bird vs when it's playing an idle animation during the ready phase.

I renamed it to `isIdle` to give it more context.

File: osu.Framework.Templates/templates/template-flappy/FlappyDon.Game/Elements/Bird.cs (hunk 23 above)
Comment (TimOliver) [reply osu.Framework.Templates/templates/template-flappy/FlappyDon.Game/Elements/Bird.cs]: I got as far as making it an inline constructor, but I'm not sure how exactly to move the frames in there, whilst keeping the duration value.

File: osu.Framework.Templates/templates/template-flappy/FlappyDon.Game/Elements/Bird.cs (hunk 24 above)
Comment (TimOliver) [reply osu.Framework.Templates/templates/template-flappy/FlappyDon.Game/Elements/Bird.cs]: Ah yeah. The point of that was to shrink the box horizontally for the sake of collision detection. Because the collision box also rotates, hitting the inside of the top pipe happens with a very noticeable gap.

I refactored this to be a lot more obvious, and less hard coded by implementing a specific `collisionQuad` property that takes the real screen space quad and shrinks it using relative values.

File: osu.Framework.Templates/templates/template-flappy/FlappyDon.Game/FlappyDonGame.cs (hunk 27 above)
Comment (TimOliver) [reply osu.Framework.Templates/templates/template-flappy/FlappyDon.Game/FlappyDonGame.cs]: Fair enough! I had to spend a bit of time thinking how to do this, but I was able to refactor it to contain all of the logic to the `Obstacles` layer, and exposed an `Action` that forwards the events back to the main game object.

File: osu.Framework.Templates/templates/template-flappy/FlappyDon.Game/Elements/ScreenFlash.cs (hunk 36 above)
Comment (TimOliver) [reply osu.Framework.Templates/templates/template-flappy/FlappyDon.Game/Elements/ScreenFlash.cs]: Removed.

File: osu.Framework.Templates/templates/template-flappy/FlappyDon.Game/FlappyDonGame.cs (hunk 37 above)
Comment (TimOliver) [reply osu.Framework.Templates/templates/template-flappy/FlappyDon.Game/FlappyDonGame.cs]: Done. And I refactored the screen flash sprite to expose manual control of the fade in and out durations, and added it here, so it could be synchronised explicitly here too.

File: osu.Framework.Templates/templates/template-flappy/FlappyDon.Game/Elements/BackdropSprite.cs (hunk 19 above)
Comment (TimOliver) [reply osu.Framework.Templates/templates/template-flappy/FlappyDon.Game/Elements/BackdropSprite.cs]: At some point, I was getting a crash because this line of code was getting called before texture was loaded. I tested and can safely remove it now.

File: osu.Framework.Templates/templates/template-flappy/FlappyDon.Game/Elements/Obstacles.cs (hunk 26 above)
Comment (TimOliver) [reply osu.Framework.Templates/templates/template-flappy/FlappyDon.Game/Elements/Obstacles.cs]: Added a test scene.

File: osu.Framework.Templates/templates/template-flappy/FlappyDon.Game/Elements/Obstacles.cs (hunk 31 above)
Comment (TimOliver) [reply osu.Framework.Templates/templates/template-flappy/FlappyDon.Game/Elements/Obstacles.cs]: Erm, it says it can't recognise `Cast`. 😅

Comment (bdach) [reply osu.Framework.Templates/templates/template-flappy/FlappyDon.Game/Elements/Obstacles.cs]: Yeah, can't do a `Cast<T>` on a single item. If `InternalChildren` is guaranteed to have only `PipeObstacle`s you can swap order and do
```csharp
InternalChildren.Cast<PipeObstacle>().FirstOrDefault()? // and so on
```
but if it isn't you probably want
```csharp
InternalChildren.OfType<PipeObstacle>().FirstOrDefault()? // and so on
```

File: osu.Framework.Templates/templates/template-flappy/FlappyDon.Game/Elements/Obstacles.cs (hunk 32 above)
Comment (TimOliver) [reply osu.Framework.Templates/templates/template-flappy/FlappyDon.Game/Elements/Obstacles.cs]: Done.

File: osu.Framework.Templates/templates/template-flappy/FlappyDon.Game/Elements/Obstacles.cs (hunk 31 above)
Comment (TimOliver) [reply osu.Framework.Templates/templates/template-flappy/FlappyDon.Game/Elements/Obstacles.cs]: Awesome! That was perfect! Thank you!

File: osu.Framework.Templates/templates/template-flappy/FlappyDon.Game/Elements/Obstacles.cs (hunk 30 above)
Comment (TimOliver) [reply osu.Framework.Templates/templates/template-flappy/FlappyDon.Game/Elements/Obstacles.cs]: Done!

File: osu.Framework.Templates/templates/template-flappy/FlappyDon.Game/FlappyDonGame.cs (hunk 29 above)
Comment (TimOliver) [reply osu.Framework.Templates/templates/template-flappy/FlappyDon.Game/FlappyDonGame.cs]: Aha. That's right. I did that because I felt the methods were different enough (ie, they return their own base implementations) that trying to create a generic one might be more complexity.

I had a play with it and was able to collapse it down to one method. :)

File: osu.Framework.Templates/templates/template-flappy/FlappyDon.Game/FlappyDonGame.cs (hunk 28 above)
Comment (TimOliver) [reply osu.Framework.Templates/templates/template-flappy/FlappyDon.Game/FlappyDonGame.cs]: Done.

**Comment by TimOliver** [comment]
Alrighty. Sorry for the delay! Thanks for all of the feedback from everyone! I've gone through and implemented all of the requested changes as best as I can. Please let me know what you think!

**Comment by peppy** [comment]
I've done a final tidying pass. @TimOliver plase have a check over the changes I made for future contribs. There's probably more that can be done later but I don't think it should delay getting this merged.
